src/Fixer/ClassNotation/FinalInternalClassFixer.php

```diff
@@ -195,23 +237,41 @@ private function isClassCandidate(TokensAnalyzer $tokensAnalyzer, Tokens $tokens
             return false; // ignore class; it is abstract or already final
         }
 
-        $docToken = $tokens[$tokens->getPrevNonWhitespace($index)];
+        $index = $tokens->getPrevNonWhitespace($index);
+
+        if ($this->checkAttributes && $tokens[$index]->isGivenKind(CT::T_ATTRIBUTE_CLOSE)) {
+            if (!$this->isClassCandidateBasedOnAttribute($tokens, $index)) {
+                return false;
+            }
+
+            while ($tokens[$index]->isGivenKind(CT::T_ATTRIBUTE_CLOSE)) {
```

Comment:
Is it possible to refactor it in a way, where `while ($tokens[$attributeCloseIndex]->isGivenKind(CT::T_ATTRIBUTE_CLOSE))` is looped only once? Because `isClassCandidateBasedOnAttribute()` internally also does the same `while`.

Maybe this method should return `?int` instead, and `return false;` should be done on returned `null`, and `int` should be used as `$index`? Thinking out loud 🙂. I'm just wondering if we need 2 similar loops.

Reply:
I agree it wasn't easy to read. Looking at the code we don't actually need the index in the return, so I've refactored the double-loop out.